### PR TITLE
armv7a: Use better set/longjmp implementation from armv7m 

### PR DESCRIPTION
**Pull request by agkaminski**
DONE: RTOS-600

## Motivation and Context

Fixes https://github.com/phoenix-rtos/libphoenix/issues/175

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [x] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [ ] Tested by hand on: (list targets here).

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: arch/armv7a/jmp.S [hunk 1]
```diff
@@ -6,90 +6,107 @@
...
  */
 
+ /* setjmp env buffer description
+  * Offset	Description
+  * 0		Signal mask flag - if != 0 then next value stores a signal mask
+  * 4		Signal mask
+  * 8		SP register
+  * 12-44	r4-r11 registers
+  * 44		LR register
+  * 48-80	d8-d15 registers
```
Comment (kemonats) [inline arch/armv7a/jmp.S:24]:
```suggestion
  * 12-43	r4-r11 registers
  * 44		LR register
  * 48-79	d8-d15 registers
```

File: arch/armv7a/jmp.S [hunk 2]
```diff
@@ -6,90 +6,107 @@
...
+  * 80		fpscr register
+  */
+
 #define __ASSEMBLY__
 
 .thumb
 .syntax unified
```
Comment (kemonats) [inline arch/armv7a/jmp.S:31]:
```suggestion
.thumb
.syntax unified

.text
```

File: arch/armv7a/jmp.S [hunk 3]
```diff
@@ -6,90 +6,107 @@
...
-	ldr r1, =(1f + 1)
-	mov r2, sp
-	/* store everythig in jmpbuf */
-	stm r0, {r1, r2, r4, r5, r6, r7, r8, r9, r10, r11, lr}
+
+.globl sigsetjmp
+.type sigsetjmp, %function
+sigsetjmp:
+	/* Store signal mask if savesigs != 0 */
+	mov r2, #0
```
Comment (kemonats) [inline arch/armv7a/jmp.S:38]:
```suggestion
	movs r2, #0
```
2 bytes instead 4
same in line 44 
```
movs r0, #0
or 
mov r0, r2 /* r2 contains 0 */
```

File: arch/armv7a/jmp.S [hunk 4]
```diff
@@ -6,90 +6,107 @@
...
+	cbz r1, 1f
+
+	/* Get signal mask */
+	/* Stack must be aligned to 8 bytes */
+	push {r0, r1, r3, lr}
 	mov r0, #0
```
Comment (kemonats) [inline arch/armv7a/jmp.S:44]:
```
movs r0, #0
or 
mov r0, r2 /* r2 contains 0 */
```

File: arch/armv7a/jmp.S [hunk 5]
```diff
@@ -6,90 +6,107 @@
...
+	stmia r0!, {r1-r11, lr}
+
+#ifndef __SOFTFP__
+	/* Store fpu registers */
+	vstm r0!, {d8-d15}
+	vmrs r2, fpscr
+	str r2, [r0], #4
+#endif
+
+	mov r0, #0
```
Comment (kemonats) [inline arch/armv7a/jmp.S:62]:
```suggestion
	movs r0, #0
```

File: arch/armv7a/jmp.S [hunk 6]
```diff
@@ -6,90 +6,107 @@
...
 	bx lr
+.size sigsetjmp, .-sigsetjmp
+
+.globl _setjmp
```
Comment (kemonats) [inline arch/armv7a/jmp.S:66]:
```suggestion
.size sigsetjmp, .-sigsetjmp


.globl _setjmp
```

File: arch/armv7a/jmp.S [hunk 7]
```diff
@@ -6,90 +6,107 @@
...
+.type _setjmp, %function
+_setjmp:
+	mov r1, #0
```
Comment (kemonats) [inline arch/armv7a/jmp.S:69]:
```suggestion
	movs r1, #0
```

File: arch/armv7a/jmp.S [hunk 8]
```diff
@@ -6,90 +6,107 @@
...
+	b sigsetjmp
 .size _setjmp, .-_setjmp
 
+.globl setjmp
```
Comment (kemonats) [inline arch/armv7a/jmp.S:73]:
```suggestion
.size _setjmp, .-_setjmp


.globl setjmp
```

File: arch/armv7a/jmp.S [hunk 9]
```diff
@@ -6,90 +6,107 @@
...
+.type setjmp, %function
+setjmp:
+	mov r1, #1
```
Comment (kemonats) [inline arch/armv7a/jmp.S:76]:
```suggestion
	movs r1, #1
```

**Comment by kemonats** [comment]
Two general comments:
1. You can add a `.align 2` directive before each function definition to align the function code to word size.
2. To consider defining the `jmp_buf` buffer size depending on the processor and the floating-point implementation used.
It is now defined like this:
https://github.com/phoenix-rtos/libphoenix/blob/345adf6103cb36c8fa9ef8b1c34e185759a93e75/include/setjmp.h#L25

File: arch/armv7a/jmp.S (hunk 3 above)
Comment (agkaminski) [reply arch/armv7a/jmp.S:38]: According to docs it shouldn't matter: https://developer.arm.com/documentation/ddi0308/d/Thumb-Instructions/Alphabetical-list-of-Thumb-instructions/MOV--immediate-
and I'd rather not use `s` here, as it makes the code somewhat confusing

File: arch/armv7a/jmp.S (hunk 4 above)
Comment (agkaminski) [reply arch/armv7a/jmp.S:44]: Again, I see no reason why should s produce different instruction encoding

File: arch/armv7a/jmp.S (hunk 1 above)
Comment (agkaminski) [reply arch/armv7a/jmp.S]: Fixed, I also spotted tabs used for fancy alignment, changed to spaces

**Comment by agkaminski** [comment]
Regarding general remarks:
1. It shouldn't be strictly needed, but might be beneficial from what I see, changed,
2. yeah, I've noticed that and I think that the common, largest size across all archs is good enough, at least until we get some exotic target with a huge `jmp_buff` needs

**Review by badochov — APPROVED** [review]

File: arch/armv7a/jmp.S (hunk 3 above)
Comment (kemonats) [reply arch/armv7a/jmp.S:42]: Such instructions are used by the compiler.
A simple example:
```c
typedef unsigned int jmp_buf[34];
typedef jmp_buf sigjmp_buf;

extern int sigsetjmp(sigjmp_buf env, int savesigs);

int _setjmp(jmp_buf var)
{
    return sigsetjmp(var, 0);
}
```
and disassembly
```
Disassembly of section .text._setjmp:

00000000 <_setjmp>:
   0:   2100            movs    r1, #0
   2:   b508            push    {r3, lr}
   4:   f7ff fffe       bl      0 <sigsetjmp>
   8:   bd08            pop     {r3, pc}
   a:   bf00            nop
```

File: arch/armv7a/jmp.S (hunk 4 above)
Comment (kemonats) [reply arch/armv7a/jmp.S:48]: You can use the second suggestion:
```
mov r0, r2 /* r2 contains 0 */
```

File: arch/armv7a/jmp.S (hunk 3 above)
Comment (agkaminski) [reply arch/armv7a/jmp.S:42]: Ok, I see now. As it turns out there're only two 16-bit `mov` immediate instructions:
```
movs Rd, #imm8
mov<c> Rd, #imm8
```
There're no 16-bit `mov` instruction to be used outside IT block (https://developer.arm.com/documentation/dui0489/i/arm-and-thumb-instructions/mov). It all makes sense now to me, will change.

**Review by badochov — APPROVED** [review]

**Pull request merged**